tests/integration/test_kubernetes_worker_integration.py

```diff
@@ -20,37 +29,82 @@ def _check_status_messages(ops_test):
             assert unit.workload_status_message == message
 
 
+@pytest.fixture()
+async def setup_resources(ops_test, tmpdir):
+    """Provides the cni resources needed to deploy the charm."""
+    cwd = Path.cwd()
+    current_resources = list(cwd.glob("*.tgz"))
+    if not current_resources:
+        # If they are not locally available, try to build them
+        log.info("Build Resources...")
+        build_script = cwd / "build-cni-resources.sh"
+        rc, stdout, stderr = await ops_test.run(
+            *shlex.split(f"sudo {build_script}"), cwd=tmpdir, check=False
+        )
+        if rc != 0:
+            log.warning(f"build-cni-resources failed: {(stderr or stdout).strip()}")
+        current_resources = list(Path(tmpdir).glob("*.tgz"))
+    if not current_resources:
+        # if we couldn't build them, just download a fixed version
+        log.info("Downloading Resources...")
+        await asyncio.gather(
+            *(
+                _retrieve_url(816, arch, tmpdir / f"cni-{arch}.tgz")
+                for arch in ("amd64", "arm64", "s390x")
+            )
+        )
+        current_resources = list(Path(tmpdir).glob("*.tgz"))
+
+    yield current_resources
```

Comment:
Should it clean up resources after the fact? I usually lean toward being good stewards and cleaning up after ourselves, but it's also not useful to fail a test if cleanup fails or putting much extra time / effort into cleanup which could even make the test timeout. Also, in this case, keeping them around could also save time effort on future runs.

Reply:
these are all downloaeded to a `.tox/integration/tmp` which gets deleted on the next run.  This is very similar to what `pytest_ops` does when it builds the charm or renders the bundle.  They all get cleaned on the next run.  This code doesn't attempt to reuse any other downloaded resources

Reply:
Oh, I missed that they were both using `tmpdir`.  +1